federation/pkg/kubefed/init/init.go

```diff
@@ -245,24 +246,24 @@ func (i *initFederation) Run(cmdOut io.Writer, config util.AdminConfig) error {
 	// 2. Expose a network endpoint for the federation API server
 	svc, ips, hostnames, err := createService(hostClientset, i.commonOptions.FederationSystemNamespace, serverName, i.options.apiServerAdvertiseAddress, i.options.apiServerServiceType, i.options.dryRun)
 	if err != nil {
-		return err
+		return cleanupObjects(hostClientset, i.commonOptions.FederationSystemNamespace, i.options.dryRun, err)
```

Comment:
This strategy would also allow the separation of error handling from cleanup.  Having to pass the error to ``cleanupObjects`` is a definite code smell.